drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -0,0 +1,819 @@
+ * Copyright (c) 2017, I-SENSE group of ICCS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <string.h>
+#include <usb/usb_device.h>
+#include <sys/util.h>
+#include <hardware/regs/usb.h>
+#include <hardware/structs/usb.h>
+#include <hardware/resets.h>
+
+#define LOG_LEVEL CONFIG_USB_DRIVER_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(udc_rpi, CONFIG_USB_DRIVER_LOG_LEVEL);
+
+#define DT_DRV_COMPAT raspberrypi_pico_usbd
+
+#define USB_BASE_ADDRESS DT_INST_REG_ADDR(0)
+#define USB_IRQ DT_INST_IRQ_BY_NAME(0, usbctrl, irq)
+#define USB_IRQ_PRI DT_INST_IRQ_BY_NAME(0, usbctrl, priority)
+#define USB_NUM_BIDIR_ENDPOINTS DT_INST_PROP(0, num_bidir_endpoints)
+/* Size of a USB SETUP packet */
+#define SETUP_SIZE 8
+
+/* Helper macros to make it easier to work with endpoint numbers */
+#define EP0_IDX 0
+
+#define DATA_BUFFER_SIZE 64U
+
+/* Needed for pico-sdk */
+#ifndef typeof
+#define typeof __typeof__
+#endif
+
+#define usb_hw_set hw_set_alias(usb_hw)
+#define usb_hw_clear hw_clear_alias(usb_hw)
+
+/* Endpoint state */
+struct udc_rpi_ep_state {
+	uint16_t mps; /** Endpoint max packet size */
+	enum usb_dc_ep_transfer_type type; /** Endpoint type */
+	uint8_t halted; /** Endpoint stall flag */
+	usb_dc_ep_callback cb; /** Endpoint callback function */
+	uint32_t read_offset; /** Current offset in read buffer */
+	struct k_sem write_sem; /** Write boolean semaphore */
+	io_rw_32 *ep_ctl;
+	io_rw_32 *buf_ctl;
+	uint8_t *buf;
+	uint8_t next_pid;
+};
+
+#define USBD_THREAD_STACK_SIZE		1024
+
+/* Driver state */
+struct udc_rpi_state {
+	usb_dc_status_callback status_cb; /* Status callback */
+	struct udc_rpi_ep_state out_ep_state[USB_NUM_BIDIR_ENDPOINTS];
+	struct udc_rpi_ep_state in_ep_state[USB_NUM_BIDIR_ENDPOINTS];
+	bool setup_available;
+	bool should_set_address;
+	uint8_t addr;
+
+	K_KERNEL_STACK_MEMBER(thread_stack, USBD_THREAD_STACK_SIZE);
+	struct k_thread thread;
+};
+
+static struct udc_rpi_state state;
+
+#define USB_DC_CB_TYPE_MGMT		0
+#define USB_DC_CB_TYPE_EP		1
+
+struct cb_msg {
+	uint8_t ep;
+	uint8_t type;
+	uint32_t cb;
+};
+
+K_MSGQ_DEFINE(usb_dc_msgq, sizeof(struct cb_msg), 10, 4);
+static void udc_rpi_thread_main(void *arg1, void *unused1, void *unused2);
+
+/* Internal functions */
+
+static struct udc_rpi_ep_state *udc_rpi_get_ep_state(uint8_t ep)
+{
+	struct udc_rpi_ep_state *ep_state_base;
+
+	if (USB_EP_GET_IDX(ep) >= USB_NUM_BIDIR_ENDPOINTS) {
+		return NULL;
+	}
+
+	if (USB_EP_DIR_IS_OUT(ep)) {
+		ep_state_base = state.out_ep_state;
+	} else {
+		ep_state_base = state.in_ep_state;
+	}
+
+	return ep_state_base + USB_EP_GET_IDX(ep);
+}
+
+static int udc_rpi_start_xfer(uint8_t ep, const void *data, size_t len)
+{
+	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(ep);
+
+	uint32_t val = len | USB_BUF_CTRL_AVAIL;
+
+	if (USB_EP_DIR_IS_IN(ep)) {
+		/* If writing to host, put data in place */
+
+		if (data) {
+			memcpy(ep_state->buf, data, len);
+			val |= USB_BUF_CTRL_FULL;
```

Comment:
`val |= USB_BUF_CTRL_FULL;` should go outside of the condition, otherwise, if the packet size is a multiplicity of the buffer size then the transfer will stall. This is a zero-length-packet.

Reply:
Moved this out. Thanks!